capa/features/extractors/ghidra/file.py

```diff
@@ -0,0 +1,198 @@
+            e_lfanew_bytes = e_lfanew_bytes + b
+        newoff = struct.unpack("<I", capa.features.extractors.helpers.xor_static(e_lfanew_bytes, i))[0]
+
+        # assume XOR'd "PE" bytes exist within threshold
+        if newoff > MAX_OFFSET_PE_AFTER_MZ:
+            continue
+
+        peoff = off.add(newoff)
+        if seg_max.getOffset() < (peoff.getOffset() + 2):
+            continue
+
+        if getBytes(peoff, 2) == pex:
+            yield off, i
+
+
+def extract_file_embedded_pe() -> Iterator[Tuple[Feature, Address]]:
+    """extract embedded PE features
+    """
+    for seg in currentProgram.getMemory().getBlocks():
+        for ea, _ in check_segment_for_pe(seg):
+            yield Characteristic("embedded pe"), FileOffsetAddress(ea)
+
+
+def extract_file_export_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract function exports
+
+    Exports in Ghidra:
+        - namespace = Global
+        - SourceType = IMPORTED
+    """
+
+    fm = currentProgram.getFunctionManager().getFunctions(True) # Only Global functions
+    while fm.hasNext():
+        f = fm.next().getSymbol()
+        if (f.getSource() == SourceType.IMPORTED):
+            yield Export(f.toString()), AbsoluteVirtualAddress(f.getAddress().getOffset())
+
+
+def extract_file_import_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract function imports
+
+    1. imports by ordinal:
+     - modulename.#ordinal
+
+    2. imports by name, results in two features to support importname-only
+       matching:
+     - modulename.importname
+     - importname
+    """
+
+    fm = currentProgram.getFunctionManager().getExternalFunctions()
+    while fm.hasNext():
+        f = fm.next()
+        addr = int(f.getEntryPoint().toString().split(':')[1], 16)  # format: EXTERNAL:<hex_addr>
+        fstr = f.getName()  # format: 'importname' / 'Ordinal_*'
+        if 'Ordinal_' in fstr:
+            fstr = f"#{fstr.split('_')[1]}"
+        yield Import(fstr), AbsoluteVirtualAddress(addr)
+
+
+def extract_file_section_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract section names"""
+
+    for block in currentProgram.getMemory().getBlocks():
+        yield Section(block.getName()), AbsoluteVirtualAddress(block.getStart().getOffset())
+
+
+def extract_file_strings() -> Iterator[Tuple[Feature, Address]]:
+    """extract ASCII and UTF-16 LE strings"""
+
+    dat = getFirstData()
+    while (dat != None):
+        if (dat.hasStringValue()):
+            yield String(dat.getValue()), FileOffsetAddress(dat.getAddress().getOffset()) 
+        dat = getDataAfter(dat)
+
+
+def extract_file_function_names() -> Iterator[Tuple[Feature, Address]]:
+    """
+    extract the names of statically-linked library functions.
+    """
+
+    fm = currentProgram.getFunctionManager().getExternalFunctions()
+    while fm.hasNext():
+        f = fm.next()
+        addr = int(f.getEntryPoint().toString().split(':')[1], 16)  # format: EXTERNAL:<hex_addr>
+        name = f.getName()
```

Comment:
does this only get names of identified (library) functions?

Reply:
Same concern

Reply:
From what I've seen so far, I believe it defaults to the `Ordinal_*` format. Are there any samples that come to mind that I may be able to test on?

Reply:
Thank you for catching this I've updated this to extract all function names in 37ece1237bd5e0e2a9977672df2d1f6f11b711dc

Reply:
In fd6b87750cc80f0ed4d73da7a1fee5f877ea1618 I've updated the code to pull all of the external symbols and then filter it further to only the function symbols i.e. the statically-linked library functions. For the samples I've tested on, it seems to pull the import name or ordinal. Are there any samples in the test-files that might allow me to test further?

Reply:
You manually verify the following tests pass: https://github.com/mandiant/capa/blob/33a3170bc4160dbfe925ef053e61823ece245996/tests/fixtures.py#L765-L803

Reply:
So using `getExternalSymbols()` and `getExternalFunctions()` would actually cause some misses on this sample when it came to the FunctionName extraction, so instead I've reverted to pulling all of the symbols, then filtering them by `SymbolType == SymbolType.FUNCTION and SourceType == SourceType.IMPORTED`. 
> Please see: afcbb0b958b528ab022a8478fe3cb6525be8760f

The local test now passes:
![image](https://github.com/mandiant/capa/assets/66766340/632a8820-31cd-436d-8a31-3455b508b633)

Reply:
![image](https://github.com/mandiant/capa/assets/66766340/49b5e7ea-7749-4367-a441-60cddde17382)

Test can currently identify the function name; however, for this specific sample, I found that the viv backend fails to extract it while Ghidra processes this symbol as an export.

ref: 12286312b457b34ca8773a16c199a324a205284f

Reply:
However, both backends match up well on PE files, rather than ELFs.